iml-gui/crate/src/components/command_modal.rs

```diff
@@ -284,3 +539,662 @@ fn close_button() -> Node<Msg> {
+    let mut ids: Vec<_> = ids.into_iter().map(|x| ("id__in", x)).collect();
+    ids.push(("limit", 0));
+    match Request::api_query(T::endpoint_name(), &ids) {
+        Ok(req) => req.fetch_json_data(data_to_msg).await,
+        Err(_) => {
+            // we always can url encode a vector of u32-s
+            unreachable!("Cannot encode request for {} with params {:?}", T::endpoint_name(), ids)
+        }
+    }
+}
+
+fn extract_uri_id<T: EndpointName>(input: &str) -> Option<u32> {
+    lazy_static::lazy_static! {
+        static ref RE: Regex = Regex::new(r"/api/(\w+)/(\d+)/").unwrap();
+    }
+    RE.captures(input).and_then(|cap: Captures| {
+        let s = cap.get(1).unwrap().as_str();
+        let t = cap.get(2).unwrap().as_str();
+        if s == T::endpoint_name() {
+            t.parse::<u32>().ok()
+        } else {
+            None
+        }
+    })
+}
+
+fn is_finished(cmd: &RichCommand) -> bool {
+    cmd.complete
+}
+
+fn is_all_finished(cmds: &HashMap<u32, Arc<RichCommand>>) -> bool {
+    cmds.values().all(|c| is_finished(c))
+}
+
+fn perform_click(select: &mut Select, id: TypedId) -> bool {
+    if select.contains(id) {
+        !select.0.remove(&id)
+    } else {
+        select.0.insert(id)
+    }
+}
+
+fn extract_children_from_cmd(cmd: &Command) -> (u32, Vec<u32>) {
+    let mut deps = cmd
+        .jobs
+        .iter()
+        .filter_map(|s| extract_uri_id::<Job0>(s))
+        .collect::<Vec<u32>>();
+    deps.sort();
+    (cmd.id, deps)
+}
+
+fn extract_children_from_job(job: &Job0) -> (u32, Vec<u32>) {
+    let mut deps = job
+        .steps
+        .iter()
+        .filter_map(|s| extract_uri_id::<Step>(s))
+        .collect::<Vec<u32>>();
+    deps.sort();
+    (job.id, deps)
+}
+
+const fn extract_children_from_step(step: &Step) -> (u32, Vec<u32>) {
+    (step.id, Vec::new()) // steps have no descendants
+}
+
+fn extract_wait_fors_from_job(job: &Job0, jobs: &HashMap<u32, Arc<RichJob>>) -> (u32, Vec<u32>) {
+    // Extract the interdependencies between jobs.
+    // See [command_modal::tests::test_jobs_ordering]
+    let mut deps = job
+        .wait_for
+        .iter()
+        .filter_map(|s| extract_uri_id::<Job0>(s))
+        .collect::<Vec<u32>>();
+    deps.sort_by(|i1, i2| {
+        let t1 = jobs
+            .get(i1)
+            .map(|arj| (-(arj.deps.len() as i32), &arj.description[..], arj.id))
+            .unwrap_or((0, "", *i1));
+        let t2 = jobs
+            .get(i2)
+            .map(|arj| (-(arj.deps.len() as i32), &arj.description[..], arj.id))
+            .unwrap_or((0, "", *i2));
+        t1.cmp(&t2)
+    });
+    (job.id, deps)
+}
+
+fn extract_sorted_keys<T>(hm: &HashMap<u32, T>) -> Vec<u32> {
+    let mut ids = hm.keys().copied().collect::<Vec<_>>();
+    ids.sort();
+    ids
+}
+
+fn convert_to_sorted_vec<T: Clone>(hm: &HashMap<u32, T>) -> Vec<T> {
+    extract_sorted_keys(hm).into_iter().map(|k| hm[&k].clone()).collect()
+}
```

Comment:
```suggestion
fn convert_to_sorted_vec<T: Clone>(hm: &HashMap<u32, Arc<T>>) -> Vec<Arc<T>> {
    extract_sorted_keys(hm)
        .into_iter()
        .map(|k| Arc::clone(&hm[&k]))
        .collect()
}
```

Nit, this would make it clearer that the clone is a ref count bump instead of a deep clone (This still compiles btw).